## 1) Initialisation & Chargement des données

In [177]:
import pyspark
print(pyspark.__version__)

3.5.3


In [178]:
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, lit, datediff, max as spark_max, sum as spark_sum, countDistinct, year, month, dayofweek, hour, quarter, count, round as spark_round, avg as spark_avg, when, desc


In [179]:
spark = SparkSession.builder.appName("Spark_Final_Project").master("local[*]").getOrCreate()

df_raw = spark.read.option("Header", True).option("inferSchema", True).option("delimiter", ";").csv("../data/raw/Online_Retail_CSV.csv")

df_raw.printSchema()
df_raw.show(5)

root
 |-- InvoiceNo: string (nullable = true)
 |-- StockCode: string (nullable = true)
 |-- Description: string (nullable = true)
 |-- Quantity: integer (nullable = true)
 |-- InvoiceDate: string (nullable = true)
 |-- UnitPrice: string (nullable = true)
 |-- CustomerID: integer (nullable = true)
 |-- Country: string (nullable = true)

+---------+---------+--------------------+--------+----------------+---------+----------+--------------+
|InvoiceNo|StockCode|         Description|Quantity|     InvoiceDate|UnitPrice|CustomerID|       Country|
+---------+---------+--------------------+--------+----------------+---------+----------+--------------+
|   536365|   85123A|WHITE HANGING HEA...|       6|01/12/2010 08:26|     2,55|     17850|United Kingdom|
|   536365|    71053| WHITE METAL LANTERN|       6|01/12/2010 08:26|     3,39|     17850|United Kingdom|
|   536365|   84406B|CREAM CUPID HEART...|       8|01/12/2010 08:26|     2,75|     17850|United Kingdom|
|   536365|   84029G|KNITTED UNI

In [180]:
print(f"Il y a {df_raw.count()} lignes.")
print(f"Il y a {len(df_raw.columns)} colonnes.")

Il y a 541909 lignes.
Il y a 8 colonnes.


## 2) Exploration & Prétraitement

## A) Analyse Descriptive

Tout d'abord, regardons combien il y a de clients référencés dans cette base de données :

In [181]:
print(f"Il y a {df_raw.select('CustomerID').distinct().count()} clients.")

Il y a 4373 clients.


Maintenant, regardons le nombre total de transactions :

In [182]:
transaction_nb = df_raw.select('InvoiceNo').dropDuplicates().count()
print(f"Il y a {transaction_nb} transactions uniques.")

Il y a 25900 transactions uniques.


Regardons de plus près la distribution de la variable Quantity :

In [183]:
df_raw.select("Quantity").describe().show()

+-------+------------------+
|summary|          Quantity|
+-------+------------------+
|  count|            541909|
|   mean|  9.55224954743324|
| stddev|218.08115785023406|
|    min|            -80995|
|    max|             80995|
+-------+------------------+



Nous observons qu’en moyenne, une ligne de commande contient environ 9 produits.  
Cependant, l’écart-type est relativement élevé (218), ce qui indique une forte dispersion des valeurs et la présence d’outliers dans la variable Quantity.  
En effet, certaines transactions présentent des quantités extrêmement faibles ou élevées, allant jusqu’à –80 995 et +80 995 unités.  
Ces valeurs extrêmes contribuent fortement à l’augmentation de l’écart-type et sont susceptibles de correspondre à des retours de produits, des annulations de commandes ou des commandes en gros.

Regardons egalement la distribution de la variable UnitPrice :

In [184]:
df_raw.select("UnitPrice").describe().show()

+-------+------------------+
|summary|         UnitPrice|
+-------+------------------+
|  count|            541909|
|   mean|29.921163668665333|
| stddev| 595.7455525989114|
|    min|         -11062,06|
|    max|             99,96|
+-------+------------------+



Nous observons que le prix unitaire moyen d’un produit est d’environ 30 unités monétaires.  
Toutefois, l’écart-type est particulièrement élevé (596), ce qui traduit une forte dispersion des valeurs et la présence d’outliers dans la variable UnitPrice.  
En effet, certaines transactions présentent des valeurs négatives, pouvant atteindre –11 062, qui ne correspondent pas à des prix réels mais sont probablement liées à des remboursements, des annulations de commandes ou des écritures de correction comptable.  
Ces valeurs extrêmes contribuent fortement à l’augmentation de l’écart-type.

Maintenant, intéressons-nous au nombre de pays clients :

In [185]:
client_countries = df_raw.select("Country").dropDuplicates().count()
print(f"Il y a {client_countries} pays qui sont clients.")

Il y a 38 pays qui sont clients.


Regardons le nombre de commandes par pays : 

In [186]:
df_raw.select("InvoiceNo", "Country").dropDuplicates().groupby("Country").count().orderBy("count", ascending=False).show()

+---------------+-----+
|        Country|count|
+---------------+-----+
| United Kingdom|23494|
|        Germany|  603|
|         France|  461|
|           EIRE|  360|
|        Belgium|  119|
|          Spain|  105|
|    Netherlands|  101|
|    Switzerland|   74|
|       Portugal|   71|
|      Australia|   69|
|          Italy|   55|
|        Finland|   48|
|         Sweden|   46|
|         Norway|   40|
|Channel Islands|   33|
|          Japan|   28|
|         Poland|   24|
|        Denmark|   21|
|         Cyprus|   20|
|        Austria|   19|
+---------------+-----+
only showing top 20 rows



Nous observons que le jeu de données comprend des clients provenant de 38 pays différents, ce qui indique une base de clientèle internationale.  
Toutefois, une très grande proportion des transactions est réalisée au Royaume-Uni, avec plus de 22 000 factures uniques, tandis que le deuxième pays, l’Allemagne, ne compte que 603 transactions.  
Cela met en évidence un fort déséquilibre géographique, le Royaume-Uni dominant largement le jeu de données par rapport aux autres pays.

Regardons maintenant le nombre total de commandes par clients, cela nous permettra de distinguer les clients habituels des clients ponctuels : 

In [187]:
df_raw.select("CustomerID", "InvoiceNo").dropDuplicates().groupBy("CustomerID").count().orderBy("count", ascending=False).show()

+----------+-----+
|CustomerID|count|
+----------+-----+
|      NULL| 3710|
|     14911|  248|
|     12748|  224|
|     17841|  169|
|     14606|  128|
|     15311|  118|
|     13089|  118|
|     12971|   89|
|     14527|   86|
|     13408|   81|
|     14646|   77|
|     16029|   76|
|     16422|   75|
|     14156|   66|
|     13798|   63|
|     18102|   62|
|     13694|   60|
|     15061|   55|
|     17450|   55|
|     16013|   54|
+----------+-----+
only showing top 20 rows



On remarque la présence d’un groupe CustomerID = NULL avec un nombre élevé de commandes. Cette observation suggère la présence de transactions sans identification client, ce qui motive l’analyse des valeurs manquantes présentée dans la section suivante.

## B) Vérifications de valeurs manquantes

Avant toute analyse plus poussée, il est important de vérifier les valeurs nulles :  

In [188]:
from pyspark.sql.functions import col, sum, when

df_raw.select([sum(when(col(c).isNull(), 1 ).otherwise(0)).alias(c) for c in df_raw.columns]).show()

+---------+---------+-----------+--------+-----------+---------+----------+-------+
|InvoiceNo|StockCode|Description|Quantity|InvoiceDate|UnitPrice|CustomerID|Country|
+---------+---------+-----------+--------+-----------+---------+----------+-------+
|        0|        0|       1454|       0|          0|        0|    135080|      0|
+---------+---------+-----------+--------+-----------+---------+----------+-------+



Nous observons la présence de valeurs manquantes dans deux colonnes sur huit : Description et CustomerID.  
La colonne Description contient un nombre relativement limité de valeurs manquantes (1 454), ce qui reste marginal au regard de la taille du jeu de données et n’impactera pas significativement l’analyse, d’autant plus que cette variable ne sera pas utilisée directement dans les modèles.  
En revanche, la colonne CustomerID présente un nombre important de valeurs manquantes (135 080), ce qui constitue un enjeu majeur pour l’analyse, notamment pour la segmentation client. Ces lignes devront faire l’objet d’un traitement spécifique lors de la phase de nettoyage des données.

Pour pallier au problème de valeurs manquantes mais egalement de prix et de quantité négatives, nous allons créer un nouveau dataset filtré sur ces trois conditions :  
1 - Les valeurs de CustomerID doivent êtres Non Nulles  
2 - Les valeurs de Quantity doivent êtres > 0  
3 - Les valeurs unitaires des produits doivent êtres > 0  

Nous créons donc ce dataset :

# Nettoyage des données

In [189]:
df_cleanV1 = (
    df_raw
    .filter(col("CustomerID").isNotNull())
)

La colonne InvoiceDate est initialement stockée sous forme de chaîne de caractères.  
Afin de permettre les calculs temporels nécessaires à l’analyse RFM, cette variable est convertie en format timestamp.

In [190]:
from pyspark.sql.functions import to_timestamp, col

df_cleanV1 = df_cleanV1.withColumn("InvoiceDate", to_timestamp(col("InvoiceDate"), "dd/MM/yyyy HH:mm"))
df_cleanV1.printSchema()
df_cleanV1.select("InvoiceDate").show(5)

root
 |-- InvoiceNo: string (nullable = true)
 |-- StockCode: string (nullable = true)
 |-- Description: string (nullable = true)
 |-- Quantity: integer (nullable = true)
 |-- InvoiceDate: timestamp (nullable = true)
 |-- UnitPrice: string (nullable = true)
 |-- CustomerID: integer (nullable = true)
 |-- Country: string (nullable = true)

+-------------------+
|        InvoiceDate|
+-------------------+
|2010-12-01 08:26:00|
|2010-12-01 08:26:00|
|2010-12-01 08:26:00|
|2010-12-01 08:26:00|
|2010-12-01 08:26:00|
+-------------------+
only showing top 5 rows



In [191]:
from pyspark.sql.functions import regexp_replace

df_cleanV1 = df_cleanV1.withColumn("UnitPrice", regexp_replace(col("UnitPrice"), ",", "." ).cast("double"))

df_cleanV1.printSchema()

root
 |-- InvoiceNo: string (nullable = true)
 |-- StockCode: string (nullable = true)
 |-- Description: string (nullable = true)
 |-- Quantity: integer (nullable = true)
 |-- InvoiceDate: timestamp (nullable = true)
 |-- UnitPrice: double (nullable = true)
 |-- CustomerID: integer (nullable = true)
 |-- Country: string (nullable = true)



## Ajout de Colonnes

Création Colonne TotalAmount

In [192]:
df_cleanV1 = df_cleanV1.withColumn("TotalAmount", col("Quantity") * col("UnitPrice"))

print("✓ Nouvelle colonne créée : TotalAmount = Quantity × UnitPrice")
print("\n--- Aperçu du dataset enrichi ---")
df_cleanV1.select("InvoiceNo", "CustomerID", "Quantity", "UnitPrice", "TotalAmount", "InvoiceDate").show(10)


✓ Nouvelle colonne créée : TotalAmount = Quantity × UnitPrice

--- Aperçu du dataset enrichi ---
+---------+----------+--------+---------+------------------+-------------------+
|InvoiceNo|CustomerID|Quantity|UnitPrice|       TotalAmount|        InvoiceDate|
+---------+----------+--------+---------+------------------+-------------------+
|   536365|     17850|       6|     2.55|15.299999999999999|2010-12-01 08:26:00|
|   536365|     17850|       6|     3.39|             20.34|2010-12-01 08:26:00|
|   536365|     17850|       8|     2.75|              22.0|2010-12-01 08:26:00|
|   536365|     17850|       6|     3.39|             20.34|2010-12-01 08:26:00|
|   536365|     17850|       6|     3.39|             20.34|2010-12-01 08:26:00|
|   536365|     17850|       2|     7.65|              15.3|2010-12-01 08:26:00|
|   536365|     17850|       6|     4.25|              25.5|2010-12-01 08:26:00|
|   536366|     17850|       6|     1.85|11.100000000000001|2010-12-01 08:28:00|
|   536366| 

Statistiques descriptives du dataset nettoyé

In [193]:
print("\n--- Statistiques descriptives après nettoyage ---")
df_cleanV1.select("Quantity", "UnitPrice", "TotalAmount").describe().show()


--- Statistiques descriptives après nettoyage ---
+-------+------------------+------------------+------------------+
|summary|          Quantity|         UnitPrice|       TotalAmount|
+-------+------------------+------------------+------------------+
|  count|            406829|            406829|            406829|
|   mean| 12.06130339774205|3.4604710185357925|20.401853884556093|
| stddev|248.69337001882553|  69.3151617232146| 427.5917175512857|
|    min|            -80995|               0.0|         -168469.6|
|    max|             80995|           38970.0|          168469.6|
+-------+------------------+------------------+------------------+



## Colonnes Type de transaction et NetAmount

In [ ]:
from pyspark.sql.functions import col, when, sum as spark_sum, countDistinct, datediff, lit, max as spark_max

# Ajouter la colonne NetAmount et TypeTransaction
df_cleanV2 = df_cleanV1.withColumn(
    "TypeTransaction",
    when(col("TotalAmount") >= 0, "Achat").otherwise("Remboursement")
).withColumn(
    "NetAmount",
    when(col("TypeTransaction") == "Achat", col("TotalAmount"))
    .otherwise(-col("TotalAmount"))  # remboursements sont soustraits
)

df_cleanV2.select("InvoiceNo", "CustomerID", "TotalAmount", "TypeTransaction", "NetAmount").show(10)


+---------+----------+------------------+---------------+------------------+
|InvoiceNo|CustomerID|       TotalAmount|TypeTransaction|         NetAmount|
+---------+----------+------------------+---------------+------------------+
|   536365|     17850|15.299999999999999|          Achat|15.299999999999999|
|   536365|     17850|             20.34|          Achat|             20.34|
|   536365|     17850|              22.0|          Achat|              22.0|
|   536365|     17850|             20.34|          Achat|             20.34|
|   536365|     17850|             20.34|          Achat|             20.34|
|   536365|     17850|              15.3|          Achat|              15.3|
|   536365|     17850|              25.5|          Achat|              25.5|
|   536366|     17850|11.100000000000001|          Achat|11.100000000000001|
|   536366|     17850|11.100000000000001|          Achat|11.100000000000001|
|   536367|     13047|             54.08|          Achat|             54.08|

## Colonnes temporelles


Pour identifier les patterns d'achat (clients qui achètent plutôt le weekend, en soirée, etc.)

In [196]:
# Extraire les composantes de la date
df_cleanV1 = df_cleanV1.withColumn("Year", year("InvoiceDate"))
df_cleanV1 = df_cleanV1.withColumn("Month", month("InvoiceDate"))
df_cleanV1 = df_cleanV1.withColumn("Quarter", quarter("InvoiceDate"))  # Trimestre (1-4)
df_cleanV1 = df_cleanV1.withColumn("DayOfWeek", dayofweek("InvoiceDate"))  # 1=Dimanche, 7=Samedi
df_cleanV1 = df_cleanV1.withColumn("Hour", hour("InvoiceDate"))

# Weekend ou semaine ?
df_cleanV1 = df_cleanV1.withColumn(
    "IsWeekend",
    when((col("DayOfWeek") == 1) | (col("DayOfWeek") == 7), 1).otherwise(0)
)

# Période de la journée
df_cleanV1 = df_cleanV1.withColumn(
    "TimeOfDay",
    when((col("Hour") >= 6) & (col("Hour") < 12), "Morning")
    .when((col("Hour") >= 12) & (col("Hour") < 18), "Afternoon")
    .when((col("Hour") >= 18) & (col("Hour") < 22), "Evening")
    .otherwise("Night")
)

In [197]:
# Analyses colonnes temporelles
print("\n--- Par période de la journée")
df_cleanV1.groupBy("TimeOfDay").agg(
    count("*").alias("Nb"),
    spark_round(spark_sum("TotalAmount"), 0).alias("Revenu"),
    spark_round(spark_avg("TotalAmount"), 2).alias("Moyen")
).show()

print("\n--- Weekend vs Semaine ---")
df_cleanV1.groupBy(when(col("IsWeekend") == 1, "Weekend").otherwise("Semaine").alias("Type")).agg(
    count("*").alias("Nb"),
    spark_round(spark_sum("TotalAmount"), 0).alias("Revenu"),
    spark_round(spark_avg("TotalAmount"), 2).alias("Moyenne")
).show()

print("\n--- Par Mois ---")
df_cleanV1.groupBy("Month").agg(
    spark_round(spark_sum("TotalAmount"), 0).alias("Revenu"),
    spark_round(spark_avg("TotalAmount"), 2).alias("Moyenne")
).orderBy("Month").show()


--- Par période de la journée
+---------+------+---------+-----+
|TimeOfDay|    Nb|   Revenu|Moyen|
+---------+------+---------+-----+
|  Evening|  7519| 161443.0|21.47|
|  Morning|120880|3170926.0|26.23|
|Afternoon|278430|4967697.0|17.84|
+---------+------+---------+-----+


--- Weekend vs Semaine ---
+-------+------+---------+-------+
|   Type|    Nb|   Revenu|Moyenne|
+-------+------+---------+-------+
|Semaine|343592|7515647.0|  21.87|
|Weekend| 63237| 784419.0|   12.4|
+-------+------+---------+-------+


--- Par Mois ---
+-----+---------+-------+
|Month|   Revenu|Moyenne|
+-----+---------+-------+
|    1| 475074.0|  21.68|
|    2| 436546.0|  21.44|
|    3| 579965.0|  20.85|
|    4| 426048.0|  18.37|
|    5| 648251.0|  22.42|
|    6| 608013.0|  21.84|
|    7| 574238.0|  20.88|
|    8| 616368.0|  22.28|
|    9| 931440.0|  22.82|
|   10| 974604.0|  19.22|
|   11|1132408.0|  17.26|
|   12| 897110.0|  20.15|
+-----+---------+-------+



# 3) SEGMENTATION CLIENT (NON SUPERVISÉ)

 --- 1. CRÉATION DES VARIABLES RFM ---

Calculer la date de référence (date maximale dans le dataset)

In [198]:
reference_date = df_cleanV1.select(spark_max("InvoiceDate")).first()[0]
print(f"Date de référence pour Recency : {reference_date}")

Date de référence pour Recency : 2011-12-09 12:50:00


Calculer les variables RFM par client

In [208]:
# Calcul RFM par client
rfm_data = df_cleanV2.groupBy("CustomerID").agg(
    datediff(lit(reference_date), spark_max("InvoiceDate")).alias("Recency"),
    countDistinct("InvoiceNo").alias("Frequency"),
    spark_sum("NetAmount").alias("Monetary")  # Utiliser NetAmount ici 
)

# Vérifier le résultat
rfm_data.show(10)
rfm_data.describe().show()

+----------+-------+---------+------------------+
|CustomerID|Recency|Frequency|          Monetary|
+----------+-------+---------+------------------+
|     16503|    106|        5|1442.4299999999998|
|     17420|     50|        3| 598.8299999999999|
|     15727|     16|        7|           5178.96|
|     16861|     59|        3|            195.87|
|     17389|      0|       43|32367.280000000002|
|     13623|     30|        7| 823.1200000000001|
|     13285|     23|        4|2709.1199999999994|
|     15447|    330|        1|            155.17|
|     14450|    180|        3|            483.25|
|     14570|    280|        2|218.05999999999997|
+----------+-------+---------+------------------+
only showing top 10 rows

+-------+------------------+------------------+-----------------+------------------+
|summary|        CustomerID|           Recency|        Frequency|          Monetary|
+-------+------------------+------------------+-----------------+------------------+
|  count|          

### 2. Verification du Dataframe RFM

In [211]:
rfm_data.select([count(when(col(c).isNull(), 1)).alias(c) for c in rfm_data.columns]).show()


+----------+-------+---------+--------+
|CustomerID|Recency|Frequency|Monetary|
+----------+-------+---------+--------+
|         0|      0|        0|       0|
+----------+-------+---------+--------+



### 3. ASSEMBLAGE ET STANDARDISATION DES FEATURES

In [202]:
from pyspark.ml.feature import VectorAssembler, StandardScaler


In [212]:
# Assemblage des variables RFM
assembler = VectorAssembler(
    inputCols=["Recency", "Frequency", "Monetary"], 
    outputCol="features_raw"
)
rfm_assembled = assembler.transform(rfm_data)

# Standardisation pour éviter que Monetary domine le clustering
scaler = StandardScaler(
    inputCol="features_raw", 
    outputCol="features", 
    withMean=True, 
    withStd=True
)
scaler_model = scaler.fit(rfm_assembled)
rfm_scaled = scaler_model.transform(rfm_assembled)

rfm_scaled.select("CustomerID", "Recency", "Frequency", "Monetary", "features").show(5, truncate=False)


+----------+-------+---------+------------------+----------------------------------------------------------------+
|CustomerID|Recency|Frequency|Monetary          |features                                                        |
+----------+-------+---------+------------------+----------------------------------------------------------------+
|16503     |106    |5        |1442.4299999999998|[0.14308321290028964,-0.008082483813860604,-0.07005087027783587]|
|17420     |50     |3        |598.8299999999999 |[-0.4126259382728494,-0.2222438125059558,-0.15037647251221353]  |
|15727     |16     |7        |5178.96           |[-0.7500207800565409,0.2060788448782346,0.28573270311810917]    |
|16861     |59     |3        |195.87            |[-0.32331553897716636,-0.2222438125059558,-0.18874537326655783] |
|17389     |0      |43       |32367.280000000002|[-0.9087948232488664,4.060982761335948,2.8745404055137653]      |
+----------+-------+---------+------------------+-------------------------------

### 4. Entrainement